text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,144 @@
+# Summary
+- Define supported protocols for services and outputs in libobs and OBS Studio
+- Outputs for streaming are registered with one or more compatible protocols and codecs
+- Only codecs compatible with the output and the service are shown
+- Audio codec can be chosen if various compatible, codec in simple mode and encoder advanced output mode.
+- Will be extended with RFC 39
+
+# Motivation
+Create a better management of outputs, encoders, and services with an approach focused on protocols.
+
+# Design
+
+## Outputs API
+Only outputs with the `OBS_OUTPUT_SERVICE` flag are concerned.
+
+Adding to `obs_output_info`:
+- `const char *protocols`: protocols separated with a semi-colon (ex: `"RTMP;RTMPS"`), required to register the output.
+- `const char *protocols_prefixes`: URL protocols prefixes separated with a semi-colon `"rtmp://;rtmps://"`, optional.
+  - Up to one URL prefix per protocol in `protocols`.
+  - Those URL prefixes must be in the same order as `protocols`.
+  - Each URL prefix must be unique to the protocol
+  - Meant to be used for protocol auto-detection on server URL.
+  - While registering if present, the number of URL prefix must be less or equal to the number of protocols.
+  - NOTE: URL prefixes (URI scheme + `://`) are prefered over URI schemes to avoid:
+    - matching links without prefixes (e.g. "rtmp.example.com").
+    - adding `://` each time that a auto-detection is done.
+
+Adding to this API these functions:
+- `const char *obs_output_get_protocols(const obs_output_t *output)`: returns protocols supported by the output.
+- `bool obs_output_is_protocol_registered(const char *protocol)`: return true if an output with the protocol is registered.
+- `const char *obs_output_get_prefix_protocol(const char *prefix)`: return the protocol bound to the prefix.
+- `bool obs_enum_output_service_types(const char *protocol, size_t idx, const char **id)`: enumerate all outputs types compatible with the given protocol.
+- `bool obs_enum_output_protocols_prefixes(size_t idx, const char **prefix)`: enumerate all registered URL prefixes.
+- `bool obs_enum_output_protocols(size_t idx, const char **prefix)`: enumerate all registered protocol.
+- `const char *obs_get_output_supported_video_codecs(const char *id)`: return compatible video codecs of the given output id.
+- `const char *obs_get_output_supported_audio_codecs(const char *id)`: return compatible audio codecs of the given output id.
+
+### About protocols
+
+- HLS relies on HTTP/HTTPS and it is not the only existing protocol in this case. This protocol has no unique URL prefixes.
+
+- Because of the deprecation of the protocol, FTL will not have a URL prefix asign to its output to not make it available through custom server in the UI.
```

Comment:
If I'm understanding this correctly, this would be a breaking change, since we currently support FTL URLs in Custom Server in the UI. See https://github.com/obsproject/obs-studio/commit/96ef45cef5baae3bf01fbb22051d6da4ee6ceae1.

Reply:
@RytoEX FTL is being deprecated/removed I believe? WebRTC satisfies all the needs of FTL (latency/easy to bridge with other RTP protocols)


If this RFC doesn’t support FTL will we have to merge WebRTC first? Or this RFC needs to support FTL.

Reply:
Per https://github.com/obsproject/obs-studio/discussions/4021:
> After 9 months of WebRTC support being available in OBS, we plan to remove FTL support (all code using FTL and all services entries for FTL ingests) from the OBS codebase, meaning the first major or minor OBS release that comes 9 months after WebRTC release will not support FTL.

That does not mean we should disable existing functionality in the lead up or interim.

Reply:
I forgot this part, will restore FTL prefix then.

Reply:
> If this RFC doesn’t support FTL will we have to merge WebRTC first? Or this RFC needs to support FTL.

If does support it, I just thought it was not the case in "Custom…" which is the case. I need to redo this part.